### DISCUSSION TITLE
Migrated to Prisma 5 from 3 - ERROR HY000 (1040): Too many connections

### DISCUSSION THREAD

**Discussion by webnoob** [prisma/prisma#22089] [Q&A]
### Question

Hi,

Using Prisma 3, I didn't see this error happening with my server. Now I'm on Prisma 5, I'm getting it.

When I examine `show full processlist;` I see a lot of (based on max_connections) connections which are sleeping. I did notice my sleep timeout wasn't set to a reasonable number in my MYSQL config but given I was having no issues on prisma 3 and now I'm seeing lots of issues on 5, I have to assume something has changed.

![image](https://github.com/prisma/prisma/assets/10404747/867a9cc6-fa1a-4373-bf96-e1d06b817c79)

How does Prisma handle the connection? Would you expect a lot of sleeping queries?

In terms of code, this is my PrismaCLientService from NestJS:

```ts
@Injectable()
export class PrismaClientService extends PrismaClient implements OnModuleInit, OnModuleDestroy {
  constructor () {
    super({
      log: ['info', /*'query',*/ 'warn']
    })
  }
  async onModuleInit () {
    await this.$connect()
  }

  async onModuleDestroy () {
    await this.$disconnect()
  }
}
```

I don't really know where to start debugging this so would appreciate some ideas.

### Information about Prisma Schema, Client Queries and Environment (optional)

- OS: Ubuntu
- Database: MYSQL

**Comment by ludralph** [comment]
Hi @webnoob 👋 

Thank you for raising this question.

> How does Prisma handle the connection? Would you expect a lot of sleeping queries?

Prisma manages a [connection pool](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/connection-pool) to the database, and the size of this pool can be influenced by the number of physical CPUs available. The default formula used for connection pool size is `num_physical_cpus * 2 + 1`. If you have not set a `connection_limit` in your database connection URL, Prisma will use this formula to determine the number of connections. This could potentially lead to a large number of connections if your server has many CPUs. Prisma Client does not automatically close connections that are no longer in use; they remain in the pool as idle connections. If your application opens new connections faster than the old ones can be reused, you might end up with many idle connections.

Can you please enable [Metrics](https://www.prisma.io/docs/concepts/components/prisma-client/metrics) preview feature? This allows you to monitor how Prisma interacts with your database. By enabling metrics, you can get detailed information about the state of Prisma and the database connections, which can help with diagnosing issues related to connection management.

**Reply by webnoob** [reply]
Great information - thank you. I've got a connection limit set (now) to 75 on the connection string. Something I didn't take into account is I'm using PM2 Clusters so it spins up 4 instances of my API which would end up quadrupling the amount of connections it allows. I've set the connection string to 75 and MYSQL to 300 so allow for this.

I'll enable the metrics and see what information we can glean and report back.

**Reply by webnoob** [reply]
@ludralph A question regarding this point:

> they remain in the pool as idle connections

Does this mean they would be in the mysql process list as a sleeping connection and it just gets picked up and used when it's needed? If so, I should (based on a 75 * 4 connection limit) have 300 sleeping connections and prisma will just wake one up for use. Is that how the pooling manages the connection?

The reason I ask is because I've turned my wait time on clearing the sleeping connections right down. If the sleeping connections are just the pooled connections prisma is using then I'm fighting against prisma with this and forcing prisma to keep creating connections and not pool properly.

**Reply by janpio** [reply]
> Does this mean they would be in the mysql process list as a sleeping connection and it just gets picked up and used when it's needed? If so, I should (based on a 75 * 4 connection limit) have 300 sleeping connections and prisma will just wake one up for use. Is that how the pooling manages the connection?

Yes, more or less. Of course Prisma does not initally open 75 connections in each app - only if those are actually needed.
Prisma should be able to handle connections going away - that is pretty normal.

Do you still have "Too many connections" now you have tuned your configuration?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
